### ISSUE TITLE
Google tests are not independent.

### ISSUE THREAD

**Issue by JodiTheTigger** [RichieSams/FiberTaskingLib#13]
Hi,

I tried running ```test/FiberTaskingLib-test ``` and it segfaulted :-(
```
> test/FiberTaskingLib-test
[==========] Running 7 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from FunctionalTests
[ RUN      ] FunctionalTests.CalcTriangleNum
fish: “test/FiberTaskingLib-test” terminated by signal SIGSEGV (Address boundary error)
```

However if I ran the test as ```test/FiberTaskingLib-test --gtest_shuffle``` then it worked fine.
```
> test/FiberTaskingLib-test --gtest_shuffle
Note: Randomizing tests' orders with a seed of 1607 .
[==========] Running 7 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 2 tests from TLSAbstraction
[ RUN      ] TLSAbstraction.GetSet
[       OK ] TLSAbstraction.GetSet (9 ms)
[ RUN      ] TLSAbstraction.MultipleThreadGetSet
[       OK ] TLSAbstraction.MultipleThreadGetSet (227 ms)
[----------] 2 tests from TLSAbstraction (236 ms total)

[----------] 2 tests from FiberAbstraction
[ RUN      ] FiberAbstraction.SingleFiberSwitch
[       OK ] FiberAbstraction.SingleFiberSwitch (0 ms)
[ RUN      ] FiberAbstraction.NestedFiberSwitch
[       OK ] FiberAbstraction.NestedFiberSwitch (0 ms)
[----------] 2 tests from FiberAbstraction (1 ms total)

[----------] 3 tests from FunctionalTests
[ RUN      ] FunctionalTests.ProducerConsumer
[       OK ] FunctionalTests.ProducerConsumer (1245 ms)
[ RUN      ] FunctionalTests.Maze10x10
[       OK ] FunctionalTests.Maze10x10 (8 ms)
[ RUN      ] FunctionalTests.CalcTriangleNum
[       OK ] FunctionalTests.CalcTriangleNum (118 ms)
[----------] 3 tests from FunctionalTests (1371 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 3 test cases ran. (1608 ms total)
[  PASSED  ] 7 tests.
```
So I don't know what's happening, but it seems that there is some sort of setup that's not been cleared between tests which makes the triangle test work, where it would segfault before. 

Notes:
Running on arch linux 
```Linux quaff 4.1.3-1-ARCH #1 SMP PREEMPT Wed Jul 22 20:37:12 CEST 2015 x86_64 GNU/Linux```
I built boost 1.58 from source and linked it in
Segfault was in munmap (no call stack) at the retq instruction:
```
0x7ffff6e56ae0                   b8 0b 00 00 00        mov    $0xb,%eax
0x7ffff6e56ae5  <+0x0005>        0f 05                 syscall
0x7ffff6e56ae7  <+0x0007>        48 3d 01 f0 ff ff     cmp    $0xfffffffffffff001,%rax
0x7ffff6e56aed  <+0x000d>        73 01                 jae    0x7ffff6e56af0 <munmap+16>
0x7ffff6e56aef  <+0x000f>        c3                    retq
0x7ffff6e56af0  <+0x0010>        48 8b 0d 61 83 2b 00  mov    0x2b8361(%rip),%rcx        # 0x7ffff710ee58
0x7ffff6e56af7  <+0x0017>        f7 d8                 neg    %eax
0x7ffff6e56af9  <+0x0019>        64 89 01              mov    %eax,%fs:(%rcx)
0x7ffff6e56afc  <+0x001c>        48 83 c8 ff           or     $0xffffffffffffffff,%rax
0x7ffff6e56b00  <+0x0020>        c3                    retq
```

I limited cores to 2 and still reproduced the error.
The other thread was in:
```
static inline int
__gthread_yield (void)
{
  return __gthrw_(sched_yield) ();   // <-- thread was here
}
```

other thread's call stack
```
0	sched_yield			0x7ffff6e43f47	
1	__gthread_yield	gthr-default.h	692	0x4cc9f5	
2	std::this_thread::yield	thread	270	0x4ce245	
3	FiberTaskingLib::TaskScheduler::Quit	task_scheduler.cpp	259	0x4cd9d9	
4	FunctionalTests_CalcTriangleNum_Test::TestBody	calc_triangle_num.cpp	98	0x48cd60	
5	testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>	gtest.cc	2364	0x4b9868	
6	testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>	gtest.cc	2400	0x4b3f0e	
7	testing::Test::Run	gtest.cc	2437	0x49aa60	
8	testing::TestInfo::Run	gtest.cc	2612	0x49b1d4	
9	testing::TestCase::Run	gtest.cc	2730	0x49b7f3	
10	testing::internal::UnitTestImpl::RunAllTests	gtest.cc	4604	0x4a1f09	
11	testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>	gtest.cc	2364	0x4ba6e7	
12	testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>	gtest.cc	2400	0x4b4c5e	
13	testing::UnitTest::Run	gtest.cc	4222	0x4a0c6f	
14	RUN_ALL_TESTS	gtest.h	2326	0x48d8d7	
15	main	main.cpp	16	0x48d871	
```

**Comment by RichieSams** [comment]
This is really interesting. I've been trying to figure out why the Boost.Context based builds are failing for a while now. This helps a lot.

If you have a minute, could you re-run the tests a few times using the same seed (1607) and check if the tests always pass?

**Comment by RichieSams** [comment]
Thank you for trying the library out and submitting an issue! :)

**Comment by JodiTheTigger** [comment]
I reran with both the 2 core hack, as well as the orig code, and 1607 ran fine for over 15 minutes running in a loop.

**Comment by JodiTheTigger** [comment]
I removed CalcTriangles and it passed.

But!

Then I ran --gtest_shuffle and it segfaulted in FunctionalTests.ProducerConsumer (Same place as usual).

**Comment by RichieSams** [comment]
Fixed by commits 1837d45e152e9bf2742248f127434dfad7e118d8 to 013ee8cabd4ab39226e0aeb6c492b9a025ea2cc3
